### ISSUE TITLE
incorrect lexer rule precedence with "not" rules

### ISSUE THREAD

**Issue by cvmocanu** [antlr/antlr4#2229]
Consider the following grammar:
```
grammar NotSoSimpleWrong;

print: PRINT text=NOT_NEWLINE NEWLINE;

PRINT       : 'print: ' ;
NOT_NEWLINE : ~[\r\n]+ ;
NEWLINE     : '\r'? '\n' | '\r' ;
```

It's intended to match a string such as "print: some text" with a trailing newline character.

The grammar doesn't work because the NOT_NEWLINE rule matches also the 'print: ' part.
"The Definitive ANTLR4 Reference" book says this on page 76: "ANTLR lexers resolve ambiguities between lexical rules by favoring the rule specified first."

This doesn't seem to be the case when the not operator ``~`` is used, like in this case.

For the reference, the following grammar works as expected:
```
grammar NotSoSimpleOk;

print: PRINT text=NOT_NEWLINE NEWLINE;

PRINT       : 'print: ' ;
NOT_NEWLINE : [ a-z]+ ;
NEWLINE     : '\r'? '\n' | '\r' ;
```
Of course, the ``NotSoSimpleOk`` grammar is not good enough, because I want to match any character besides newline, not just letters and space.

As a workaround, I could match "print: some text" as a single token and do substring(7) to take the text, when processing the parse tree, but that feels very dirty.

Using ANTLR version 4.7.1 from maven.

**Comment by cvmocanu** [comment]
Ok, I found a nicer workaround - lexical modes:

```
lexer grammar NotSoSimpleOkLexer;

PRINT       : 'print: ' -> pushMode(printMode);
NEWLINE     : '\r'? '\n' | '\r' ;

mode printMode;

NOT_NEWLINE : ~[\r\n]+ -> popMode;
```

```
parser grammar NotSoSimpleOkParser;

options { tokenVocab=NotSoSimpleOkLexer; }

print: PRINT text=NOT_NEWLINE NEWLINE;
```

This way, we avoid matching NOT_NEWLINE  because it's in a different lexical mode.

**Comment by cvmocanu** [comment]
It would be nice for a maintainer to check if this is indeed a bug, and if not, explain why.

**Comment by ericvergnaud** [comment]
Hi,
please discuss this in the google discussion group.
Eric

**Comment by cvmocanu** [comment]
@ericvergnaud : That is the first thing I did: https://groups.google.com/forum/#!topic/antlr-discussion/VQym0tOFXGI

Since there doesn't seem to be any difference in the lexer algorithm between the working and non-working grammars (just increment the character index while the condition remains true), I consider this a bug.
That is why I reported it here.

**Comment by KvanTTT** [comment]
This is not a bug. See this answer for example [How does the ANTLR lexer disambiguate its rules (or why does my parser produce “mismatched input” errors)?](https://stackoverflow.com/a/46267981/1046374).

**Comment by cvmocanu** [comment]
@KvanTTT: thanks for your pointer.
So the rule really is "pick the lexer rule that matches the most input".

Context-independent lexing really makes things much more difficult than they need to be.

I recently watched a [YouTube presentation](https://www.youtube.com/watch?v=q8p1voEiu8Q) by @parrt in which he said ANTLR 4 is really where he wanted a parser generator to be.
In the same presentation he complained about the wiki grammar being so difficult to parse because it's so context-dependent. May I suggest that there is nothing wrong with that. What makes is so hard to parse with ANTLR is a limitation in ANTLR, not in the grammar. It's the tool that makes things complicated, not the grammar. Context-independent lexing is a needless limitation.

As a side note, I recently parsed a grammar I created using a parser combinator library called JParsec, and it has none of the limitations of context-independent lexing. It's really a pleasure to work with it. Not to mention that it makes for really maintainable, modular code.
The only problem I have with it (and the reason why I am looking at ANTLR) is that JParsec looses the indexes (& line/column) into the input string, and for my purposes I really need those in the parse tree.

Looking forward to an ANTLR 5 that has context-dependent lexing (or removal of lexing phase completely) :)

**Comment by parrt** [comment]
just use characters not tokens in the parser and it'll do exactly that. i've done multiple times myself.

**Comment by cvmocanu** [comment]
This works, indeed:
```
grammar NotSoSimpleWrong;

print: PRINT printMessage NEWLINE;

printMessage: NOT_NEWLINE+ ;

PRINT       : 'print: ' ;
NOT_NEWLINE : ~'\n' ;
NEWLINE     : '\r'? '\n' | '\r' ;
```

I wonder if the performance impact is significant...

**Comment by parrt** [comment]
it's much slower in my experience so i try to lump what i can together into tokens and send only those char-as-tokens to parser as needed to make it work.

**Comment by parrt** [comment]
took me 30 years, but i got what I wanted ;)
